File: source/com/barrybecker4/simulation/conway/rendering/ConwayRenderer.java
// Copyright by Barry G. Becker, 2015. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.conway.rendering;

import com.barrybecker4.common.geometry.Location;
import com.barrybecker4.simulation.conway.model.ConwayProcessor;
import com.barrybecker4.ui.renderers.OfflineGraphics;
import com.barrybecker4.ui.util.ColorMap;

import java.awt.*;
import java.awt.image.BufferedImage;


/**
 * Everything we need to know to compute the l-System tree.
 * Should make the tree automatically center.
 *
 * @author Barry Becker
 */
public class ConwayRenderer {

    private static final Color FLOOR_COLOR = new Color(130, 155, 175);

    private final double width;
    private final double height;
    private final boolean showShadows;
    private final int scale;
    private ConwayProcessor processor;

    /** offline rendering is fast  */
    private final OfflineGraphics offlineGraphics_;
    private ColorMap cmap;

    /** Constructor */
    public ConwayRenderer(int width, int height, boolean showShadows, int scale,
                          ConwayProcessor processor, ColorMap cmap)
            throws IllegalArgumentException {
        this.width = width;
        this.height = height;
        this.showShadows = showShadows;
        this.scale = scale;
        this.processor = processor;
        this.cmap = cmap;
        offlineGraphics_ = new OfflineGraphics(new Dimension(width, height), FLOOR_COLOR);
    }

    public int getWidth() {
        return (int) width;
    }
    public int getHeight() {
        return (int) height;
    }

    public BufferedImage getImage() {
        return offlineGraphics_.getOfflineImage();
    }

    /** render the live cells on the grid */
    public void render() {
        for (Location c : processor.getPoints())  {
            Integer value = processor.getValue(c);
            if (value != null) {
                Color color = cmap.getColorForValue(value);
                offlineGraphics_.setColor(color);
                int xpos = c.getX() * scale;
                int ypos = c.getY() * scale;
                offlineGraphics_.fillRect(xpos, ypos, scale, scale);
            }
        }
    }
}

File: scala-source/com/barrybecker4/simulation/conway/rendering/ConwayRenderer.scala
// Copyright by Barry G. Becker, 2014-2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.conway.rendering

import com.barrybecker4.simulation.conway.model.ConwayProcessor
import com.barrybecker4.ui.renderers.OfflineGraphics
import com.barrybecker4.ui.util.ColorMap
import java.awt.{Color, Dimension}
import java.awt.image.BufferedImage

/**
  * Renders live Conway cells into a buffered image. When [[showShadows]] is true, past positions linger.
  * @author Barry Becker
  */
object ConwayRenderer {
  private val FloorColor = new Color(240, 244, 254)
}

class ConwayRenderer(
    val width: Double,
    val height: Double,
    val showShadows: Boolean,
    val scale: Int,
    var processor: ConwayProcessor,
    var cmap: ColorMap
) {

  final private val offlineGraphics: OfflineGraphics =
    new OfflineGraphics(new Dimension(width.toInt, height.toInt), ConwayRenderer.FloorColor)

  def getWidth: Int = width.toInt
  def getHeight: Int = height.toInt
  def getImage: BufferedImage = offlineGraphics.getOfflineImage.get

  /** Draw live cells using the color map (cell age). */
  def render(): Unit = synchronized {
    if !showShadows then offlineGraphics.clear()
    for c <- processor.getPoints do
      val value = processor.getValue(c)
      offlineGraphics.setColor(cmap.getColorForValue(value))
      val xpos = c.getX * scale
      val ypos = c.getY * scale
      offlineGraphics.fillRect(xpos, ypos, scale, scale)
  }
}
